cawf.1
.\"	manual page for cawf(1)
.\"
.\"
.\"	Copyright (c) 1991 Purdue University Research Foundation,
.\"	West Lafayette, Indiana 47907.  All rights reserved.
.\"
.\"	Written by Victor A. Abell <abe@cc.purdue.edu>,  Purdue
.\"	University Computing Center.  Not derived from licensed software;
.\"	derived from awf(1) by Henry Spencer of the University of Toronto.
.\"
.\"	Permission is granted to anyone to use this software for any
.\"	purpose on any computer system, and to alter it and redistribute
.\"	it freely, subject to the following restrictions:
.\"
.\"	1. The author is not responsible for any consequences of use of
.\"	   this software, even if they arise from flaws in it.
.\"
.\"	2. The origin of this software must not be misrepresented, either
.\"	   by explicit claim or by omission.  Credits must appear in the
.\"	   documentation.
.\"
.\"	3. Altered versions must be plainly marked as such, and must not
.\"	   be misrepresented as being the original software.  Credits must
.\"	   appear in the documentation.
.\"
.\"	4. This notice may not be removed or altered.
.\"
.\" Some of the stuff in this file is a bit contorted, because it's also
.\" the regression-test input.
.nr ES 5n
.de ES
.PP
.in +\\n(ESu
.nf
..
.de EE
.in -\\n(ESu
.fi
.PP
..
.de PT
.ie \\n(.$>1 .TP "\\$2"
.el .TP
.ie !'\\$1'' \\$1
.el \(bu
..
.ds Nr \fInroff\fR
.TH CAWF 1 "Version 4.08, August 7, 1995"
.BY CAWF
.SH NAME
cawf \- C version of the nroff-like, Amazingly Workable (text) Formatter
.SH SYNOPSIS
.B cawf
[
.BI \-c " config"
] [
.BI \-d " device"
] [
.B \-e
] [
.BI \-f " font"
] [
.B \-h
] [
.BI \-m acros
] [
.BI \-n " num"
] [
.BI \-o " pages"
] [ file ... ]
.SH DESCRIPTION
.if t .tm OOPS -- CAWF THINKS IT'S TROFF!!!
.I Cawf
formats the text from the input \fIfile\fR(s)
(standard input if none)
in an approximation of \*(Nr.
It comes closest to duplicating \*(Nr's
.B man
or
.B ms
macro package styles.
It has some limited support for \*(Nr's
.B me
macros.
.SH OPTIONS
Options must precede file names.
.TP
.BI \-c " config"
defines an alternate path to the device configuration file.
Normally the device configuration file is found in
.I device.cf
in the
.I cawf
library (see the
.B FILES
section).
.IP
The device configuration file contains device character strings for
selecting fonts and the bold or italic type faces.
See the
.B DEVICES
section for more information.
.TP
.BI \-d " device"
specifies the name of the output device.
There are three built\-in devices \- ANSI, NONE and NORMAL \- and
other devices may be defined in the device configuration file.
See the
.B DEVICES
section for more information.
.IP
The NORMAL device is the default.
.TP
.B \-e
directs
.I cawf
to issue an eject (FF or ^L) after the last page.
.TP
.BI \-f " font"
specifies the one font for the device, declared with the
.BI \-d device
option, that is to be used for the
entire document.
.I Font
must match a font associated with the device's stanza in the device
configuration file.
See the
.B DEVICES
section for more information.
.IP
No
.I font
may be specified for the built\-in devices ANSI, NONE or NORMAL.
.TP
.B \-h
requests a help display.
.TP
.BI \-m acro
specifies the macro file to be used.
The standard
.I cawf
distribution supplies macro files to support ``\-man'', ``\-me'' or ``\-ms''.
.I Cawf
finds a macro file by constructing its name from `m',
.I acro
and
.B .mac
\- e. g.,
.BI \-m an
is converted to
.BR man.mac .
The default directory for macro files is defined when
.I cawf
is compiled; it's \fIC:\\SYS\\LIB\\CAWF\fP in the MS\-DOS environment;
.I /Homes/abe/lib/cawf
in the UNIX environment.
.TP
.BI \-n " num"
specifies that the starting page number should be \fInum\fP.
.TP
.BI \-o " pages"
specifies pages and a range of pages to be printed in a comma
separated list.
Ranges are specified in the form ``first-last''.
If the first number of a range is omitted, it is assumed to be one.
If the last number is omitted, it is assumed to be the last page of
the document.
For example, ``-5,9,6,11-'' specifies the printing of pages one
through five, six, nine, and eleven through the end of the document.
.TP
file ...
are the names of files containing \*(Nr source text.
.SH NROFF COMPATIBILITY
.I Cawf
accepts the following raw \*(Nr requests:
.LP
	.\e"	.ad	.bp	.br	.ce	.de	.di	.ds
	.el	.fi	.fl	.ft	.i0	.ie	.if	.in
	.it	.lg	.li	.ll	.ls	.lt	.na	.ne
	.nf	.nr	.ns	.pl	.po	.ps	.rm	.rn
	.rr	.rs	.so	.sp	.ta	.ti	.tl	.tm
	.tr
.LP
and the following in-text codes:
.ES
\e$	\e%	\e*	\e"	\ec	\ef	\eh	\ek
\en	\es	\ew
.EE
plus the full list of \*(Nr/\c
.I troff
special characters in
the original V7 \fItroff\fR manual.
.PP
Many restrictions are present; the behavior in general is a subset of
\*(Nr's.  Of particular note are the following:
.IP \(bu 2
The fully supported nroff request control character is the period.
There is limited support for the  non\-break, acute accent control
character.
.PT
Point sizes do not exist;
.B .ps
is ignored.
.PT
Special vertical spacing \- the
.B .vs
request included \- is ignored.
.PT
Conditionals cover only the numeric comparisons >, =, <, >= and <= on
.BR \en(.$ ;
string com\%par\%isons between a macro parameter and a literal;
.B n
(always true);
and
.BR t
(always false).
Only single line input is accepted from conditionals;
multi\-line input \- e.g., \\(\fIanything\fP\\) \- is not supported.
.PT
The handling of strings is generally primitive.
.PT
The handling of page header and footer strings is limited.
Some in\-text code forms are supported.
When the
.B ms
macros are in use, the
.B CH
and
.B CF
strings are preset to standard
values.
(See the
.B "MS MACROS
and
.B "HEADERS AND FOOTERS
sections for more detail.)
.IP \(bu
Horizontal motion via
.B \eh
must be supplied with a number register interpolation and must be
positive - e. g.,
.BR \eh\en(NN ,
where the value in NN is >= 0.
.IP \(bu
The
.B \ek
function is reliable only after TAB characters, so it is useful only
for measuring table positions.
.IP \(bu
The
.B .di
request only turns output on and off \- any macro name is ignored.
.IP \(bu
Expressions - e. g.,
.B .sp
- are reasonably general, but the
.BR | ,
.BR & ,
and
.BR :\&
operators do not exist, there must be white space between the end of the \*(Nr
function and the beginning of the expression, and 
.B \ew
requires that quote (') be used as the delimiters.
.B \ew
counts the characters inside the quotes and scales the result in ens,
so that, for example, \ew'\e(bu' equals 4n, and \ew'\e(bu'/1n equals 4.
.PT
The only acceptable count for the
.B .it
request is one,
and it is effective only with
.BR man ,
.B me
or
.B ms
macros.
.PT
The default scaling factor is `v' for the
.BR .ne ,
.BR .sp ,
and
.B .pl
raw \*(Nr requests; it is `u' for
.BR .nr ;
and `n' for
.BR .in ,
.BR .ll ,
.BR .ls ,
.BR .lt ,
.BR .po ,
.BR .ta
and
.BR .ti .
(A different scaling factor may be specified with a trailing character.)
.PT
Some obsolete or meaningless requests \-
.BR .i0 ,
.B .lg
and
.B .li
\&\- are silently ignored.
.PT
White space at the beginning of lines,
and embedded white space within lines is dealt with properly.
.PT
Sentence terminators at ends of lines are understood to imply
extra space afterward in filled lines.
.PT
Tabs are im\%plemented crudely and not exactly, although
usually they work as expected.
.PT
Hyphenation is done only at explicit hyphens (`-' characters), em-dashes
(the \\*(em string) and \*(Nr discretionary hyphens.
.IP
The `-' character is always considered a hyphen when it is part of a word, and
is output in the Roman face, no matter what face is in effect for the word.
One or two `-' characters, preceded by a word break, will be output in the
effective face, so the
.B me
\&\\*- string will be output properly, if used as in ``word \\*-''.
.IP
If there are more than two consecutive `-' characters, and the effective
face is bold or italic, only the first two will be output in the effective
face; the remaining `-' characters will be output in the Roman face.
.PT
By default bold and italic characters are emulated with backspacing and
overprinting, but the
.B \-d
and
.B \-f
options, combined with the contents of the device configuration file,
may be used to generate special codes for bold and italic characters.
(See the
.B DEVICES
section for more information.)
.SH "MAN MACROS"
The
.B man
macro set replicates the full V7 manual macros,
plus a few semi-random oddballs.
The full list is:
.ES
\&.AT	.B	.BI	.BR	.BY	.DE	.DS	.DT	.HP	.I
\&.IB	.IP	.IR	.IX	.LP	.NB	.P	.PD	.PP	.RB
\&.RE	.RI	.RS	.SH	.SM	.SS	.TH	.TP	.UC
.EE
.B .BY
and
.B .NB
each take a single string argument (respectively, an indi\%cation of
authorship and a note about the status of the manual page) and arrange
to place it in the page footer.
.B .AT
and
.B .IX
do nothing.
.SH "ME MACROS"
The
.B me
macro subset has been derived from the
.I cawf
.B ms
macros by Chet Creider <creider@csd.uwo.ca>.
It includes:
.ES
\&.(l	.(q	.)l	.)q	.b	.bu	.i	.ip	.lp	.np
\&.pp	.r	.sh	.sm	.u	.uh
.EE
The
.B .(l
macro supports the C and L options.
Size changes via the
.B .sz
and
.B .sm
macros are silently ignored.
The
.B .u
macro just prints its argument in italics.
The
.B ps
and
.B pi 
number registers exist and can be changed.
The
.B td
string is available and is preset to today's date.
Single spacing is not enforced in lists and quotations.
.PP
The
.BR .AB ,
.BR .AE ,
.BR .AI ,
.BR .AU ,
.BR .DA ,
.BR .ND ,
.BR .TL
and
.B .UX
macros have been retained from the
.B ms
set.
The
.B .XP
macro has been borrowed from the Berkeley additions to the
.B ms
macro set.
.SH "MS MACROS"
The
.B ms
macro set is a substantial subset of the V7 manuscript macros.
The macros are:
.ES
\&.AB	.AE	.AI	.AU	.B	.CD	.DA	.DE	.DS	.I
\&.ID	.IP	.LD	.LG	.LP	.ND	.NH	.NL	.PP	.QE
\&.QP	.QS	.R	.RE	.RP	.RS	.SH	.SM	.TL	.TP
\&.UL	.UX
.EE
Size changes are recognized but ignored, as are
.B .RP
and
.BR .ND .
.B .UL
just prints its argument in italics.
.BR .DS / .DE
does not do a keep,
nor do any of the other macros that normally imply keeps.
.PP
The
.B DY
string is available, is preset to today's date, and
may be altered with the
.B .DA
macro.
The
.B MO
string is available, is preset to the current month,
and may be altered with the
.B DA
macro.
The
.BR PD ,
.BR PI ,
and
.BR LL
number registers exist and can be changed.
.PP
The
.B CH
string is preset for page numbering, and the
.B CF
string is preset to the contents of the
.B DY
string.
(The
.B DY
string can be set with the
.B .DA
macro; it is preset to today's date.)
.SH "HEADERS AND FOOTERS"
.I Cawf's
page header and footer handling is limited.
It comes closest to that of the
.B ms
macros with the following characteristics.
.IP \(bu 2
Strings may be placed in three areas \- the left, center, and right
portions of page header and footer lines.
.PT
The page header line is not normally printed on the first page of
output, but may be selected with the ``.^b fh 1'' directive.
.PT
The ``.^b lf n'' and ``.^b lh n'' directives may be used to specify
the number of lines in the page header and footer sections.
.IP
The default section length is five lines \- two blank lines, followed
by the header or footer line, followed by two blank lines.
.IP
However, the last blank footer line is suppressed, so that printers
that can't look ahead to a following form feed won't issue an extra
form feed and cause a fully blank page to be printed.
This makes the effective default footer section length four lines \-
five minus one.
.PT
The ``\\f'' (font change) in\-text code is interpreted.
Each of the three areas on the header or footer line is a separate
entity with respect to type faces.
Each begins in the Roman face and
.I cawf
closes it by returning to the Roman face, if necessary.
.PT
The ``\\n'' (number register interpolation) and ``\\*'' (string
interpolation) in\-text codes are expanded.
.PT
The `%' current page number in\-text code is interpreted.
Use the ``\\%'' two character sequence to introduce a percent
sign into a header or footer area.
.PT
Any other two character sequence, beginning with `\\', is contracted
into the second character of the sequence \- e.g., ``\\\\'' becomes
a single `\\' character.
.PT
In\-text codes within strings are not interpreted \- e.g., when a
string contains a request to interpolate a second string or change
font, the in\-text code is printed, not its interpretation.
.PP
When the
.B ms
macros are in use, the
.B CH
string is preset for page numbering, and the
.B CF
string is preset to the contents of the
.B DY
string (today's date).
(See the
.B "MS MACROS
section.)
.PP
Text is is placed in the page header and footer lines from the
.BR LH ,
.BR CH ,
.BR RF ,
.BR LF ,
.BR CF ,
and
.B RF
strings.
The
.B .^b
requests provide limited control over the printing of the page header
and footer lines:
.LP
.ta \w'.^b HF 0'u+3n
.nf
\&.^b fh 1	enables header string printing on the first page
\&.^b fh 0	disables header string printing on the first page
\&.^b HF 1	enables header/footer string printing
\&.^b HF 0	disables header/footer string printing
\&.^b lf \fIn\fP	print an (\fIn\fP - 1) line page footer (default \fIn\fP is 5)
\&.^b lh \fIn\fP	print an \fIn\fP line page header (default \fIn\fP is 5)
.fi
.LP
There are appropriate
.B .^b
requests in the distribution
.BR man ,
.B me
and
.B ms
macro files.
(The
.B me
and
.B ms
macro files use another
.B .^b
request, \fB.^b NH\fP, to enable numbered section header processing.)
.SH STRINGS
.I Cawf
defines a few string variables in addition to the standard ones
defined for nroff compatibility \- i.e.,
.BR CF ,
.BR CH ,
.BR LF ,
.BR LH ,
.BR RF ,
or
.BR RH .
These additional strings include:
.LP
.TP \w'MO'u+3n
.B DY
current date or date set via
.B .DA
(\fBms\fP macro set only)
.TP
.B IF
current input file name (STDIN for the standard input file)
.TP
.B MO

cawf.c
/*
 *	cawf - a C version of Henry Spencer's awf(1), the Amazingly
 *	       Workable (text) Formatter
 *
 *	V. Abell, Purdue University Computing Center
 */

/*
 *	Copyright (c) 1991 Purdue University Research Foundation,
 *	West Lafayette, Indiana 47907.  All rights reserved.
 *
 *	Written by Victor A. Abell <abe@cc.purdue.edu>,  Purdue	University
 *	Computing Center.  Not derived from licensed software; derived from
 *	awf(1) by Henry Spencer of the University of Toronto.
 *
 *	Permission is granted to anyone to use this software for any
 *	purpose on any computer system, and to alter it and redistribute
 *	it freely, subject to the following restrictions:
 *
 *	1. The author is not responsible for any consequences of use of
 *	   this software, even if they arise from flaws in it.
 *
 *	2. The origin of this software must not be misrepresented, either
 *	   by explicit claim or by omission.  Credits must appear in the
 *	   documentation.
 *
 *	3. Altered versions must be plainly marked as such, and must not
 *	   be misrepresented as being the original software.  Credits must
 *	   appear in the documentation.
 *
 *	4. This notice may not be removed or altered.
 */

static char Version[] = "4.10";

#include "cawf.h"

#include <time.h>

#include <sys/stat.h>

#if	!defined(UNIX)
#include <io.h>
#include <process.h>
#include <string.h>
#include <sys\types.h>
#include <sys\stat.h>
#endif	/* !defined(UNIX) */

#if	defined(macintosh)

# if	!defined(CAWFLIB)
#define	MACINTFCAWF	"Interfaces:cawf:"
# endif	/* !defined(CAWFLIB) */

extern void InitToolbox();
#endif	/* defined(macintosh) */

static char *Month[] =
	{ "January", "February", "March", "April", "May", "June", "July",
	  "August", "September", "October", "November", "December" };


main(argc, argv)
	int     argc;
	char    *argv[];
{
	char dt[32];			/* date for td and DY strings */
	char *ep;               	/* environment pointer */
	int fff = 0;			/* final form feed status */
	char *files[MAXFILES];  	/* file names */
	int help = 0;			/* help status */
	int i;	               		/* temporary index */
	size_t l;                       /* length */
	char *lib = CAWFLIB;		/* library path */
	int libl;			/* library path length */
	struct tm *lt;			/* local time structure pointer */
	int mac = 0;			/* macro specification status */
	int nf = 0;             	/* number of files */
	char *np;               	/* name pointer */
	int pc;                 	/* prolog count */
	struct stat sbuf;               /* stat buffer */
	time_t tm;			/* system time of day */
/*
 * Save program name.
 */
	if ((Pname = strrchr(argv[0], '\\')) != NULL)
		Pname++;
	else if ((Pname = strrchr(argv[0], '/')) != NULL)
		Pname++;
	else
		Pname = argv[0];

#if	defined(macintosh)
/*
 * Perform Macintosh initialization.
 */
	InitToolbox();

# if	!defined(CAWFLIB)
	if ((ep = getenv("MPW")) == NULL)
		ep = "";
	if ((CAWFLIB = (char *)malloc(strlen(ep) + strlen(MACINTFCAWF) + 1))
	== NULL) {
		(void) fprintf(stderr,
			"%s: no CAWFLIB buffer space for \"%s%s\"\n",
			Pname, ep, MACINTFCAWF);
		exit(1);
	}
	(void) sprintf(CAWFLIB, "%s%s", ep, MACINTFCAWF);
	lib = CAWFLIB;
# endif	/* !defined(CAWFLIB) */
#endif	/* defined(macintosh) */


#if	defined(__EMX__)
/*
 * Perform OS2 initialization:
 *
 *	expand response files;
 *	do file name glob'bing
 */
	_response(&argc, &argv);
	_wildcard(&argc, &argv);
#endif	/* defined(__EMX__) */

/*
 * Set error file stream pointer.
 */
	Efs = stderr;
/*
 * Get library name.
 */
	if ((np = getenv("CAWFLIB")) != NULL)
		lib = np;
	libl = strlen(lib);
/*
 * Get device file name.
 */
	for (ep = getenv("TERM");; ep = NULL) {
		if (ep == NULL || *ep == '\0')
			ep = "dumb";
		l = libl + 1 + strlen(ep) + strlen(".dev") + 1;
		if ((np = malloc(l)) == NULL)
			Error(FATAL, NOLINE,
				" no string space for device file: ", ep);
		(void) sprintf(np, "%s%s%s.dev", lib,

#if	defined(macintosh)
			"",
#else	/* !defined(macintosh) */
			"/",
#endif	/* defined(macintosh) */

			ep);
		if (stat(np, &sbuf) == 0)
			break;
		if (strcmp(ep, "dumb") == 0)
			Error(FATAL, NOLINE, " no dumb.dev file in ", lib);
		(void) free(np);
	}
	files[nf++] = np;
/*
 * Get common text file name.
 */
	l = libl + 1 + strlen("common") + 1;
	if ((np = malloc(l)) == NULL)
		Error(FATAL, NOLINE, " no string space for common file name",
			NULL);
	(void) sprintf(np, "%s%scommon", lib,

#if	defined(macintosh)
		""
#else	/* !defined(macintosh) */
		"/"
#endif	/* defined(macintosh) */

		);

	files[nf++] = np;
/*
 * Process options.
 */
	while ((i = getopt(argc, argv, "c:d:ef:hm:n:o:")) != EOF) {
		switch (i) {
	/*
	 * -c c
	 *
	 * Set the device configuration file path to c.
	 */
		case 'c':
			Devconf = optarg;
			break;
	/*
	 * -d d
	 *
	 * Set the output device name to d.
	 *
	 * The default output device name is NORMAL -- i.e., a device that
	 * does bold face with backspace and overprinting and italic face with
	 * underscore.  NORMAL is usually a terminal device.
	 *
	 * There is a built-in device, named ANSI, that does bold face with
	 * the ANSI shadow mode and italic face with the ANSI underscore mode.
	 * ANSI is normally a terminal device that supports the ANSI shadow
	 * and underscore modes.
	 *
	 * There is a built-in output device, named NONE, that does nothing
	 * at all for the bold or italic faces.  This is usually a terminal
	 * device.
	 *
	 * All other device names must match a stanza in the device
	 * configuration file.
	 */
		case 'd':
			Device = optarg;
			break;
	/*
	 * -e -- eject: issue final form feed
	 */
		case 'e':
			fff = 1;
			break;
	/*
	 * -f f
	 *
	 * Set f as the font name for the output device (from the device
	 * configuration file).
	 */
		case 'f':
			Devfont = optarg;
			break;
	/*
	 * -h -- display help (usage)
	 */
		case 'h':
			help = 1;
			break;
	/*
	 * -m m
	 *
	 * Set the macro file name to m.
	 *
	 *  Special support is provided for -man, -me and -ms.
	 */
		case 'm':
			if (mac) {
				Error(WARN, NOLINE,
					"multiple macro file declaration",
					NULL);
				break;
			}
			l = libl + 2 + strlen(optarg) + strlen(".mac") + 1;
			if ((np = malloc(l)) == NULL)
				Error(FATAL, NOLINE, " no string space for ",
					argv[1]);
			(void) sprintf(np, "%s%sm%s.mac", lib,

#if	defined(macintosh)
				"",
#else	/* !defined(macintosh) */
				"/",
#endif	/* defined(macintosh) */

				optarg);

			files[nf++] = np;
			if (strcmp(optarg, "an") == 0)
				Marg = MANMACROS;
			else if (strcmp(optarg, "e") == 0)
				Marg = MEMACROS;
			else if (strcmp(optarg, "s") == 0)
				Marg = MSMACROS;
			mac++;
			break;
	/*
	 * -n n
	 *
	 * Set the starting page number to n.
	 */
		case 'n':
			Thispg = atoi(optarg);
			break;
	/*
	 * -o o
	 *
	 * Set the pages to print list from the string o.
	 *
	 * Entries are separated by commas.  A range is specified as n-m.
	 * A starting range of the form -m has one for a starting value,
	 * and a range of the form n- has an ending value equivalent to
	 * the end of the document.  For example:
	 *
	 *	-5,8,10-
	 *
	 * specifies the printing of pages 1 through 5, 8, and 10 through
	 * the end of the document.
	 */
		case 'o':
			if (!AsmPgRange(optarg))
				Error(WARN, NOLINE, "invalid page list: ",
					optarg);
			break;
	/*
	 * Option not recognized by getopt().
	 */
		case '?':
			Err = 1;
		}
	}
	if (Defdev())
		Err++;
	if (help || Err) {
	  (void) fprintf(stderr,
	    "%s %s usage: [-c c] [-d d] [-e] [-f f] [-h] [-m m]\n",
		Pname, Version);
	  (void) fprintf(stderr, "\t[-n n] [-o o] file...\n");
	  (void) fprintf(stderr,
	    "\t-c c      c is the device configuration file path\n");
	  (void) fprintf(stderr,
	    "\t          (default = %s%s%s)\n", CAWFLIB,

#if	defined(macintosh)
	    "",
#else	/* !defined(macintosh) */
	    "/",
#endif	/* defined(macintosh) */

  	    DEVCONFIG);
	  (void) fprintf(stderr,
	    "\t-d d      d is the output device name\n");
	  (void) fprintf(stderr,
	    "\t          (default = NORMAL, using \\b for bold and italic)\n");
	  (void) fprintf(stderr,
	    "\t          (built-ins = ANSI, NONE and NORMAL)\n");
	  (void) fprintf(stderr,
	    "\t-e        issue eject after last page\n");
	  (void) fprintf(stderr,
	    "\t-f f      f is the output device font name\n");
	  (void) fprintf(stderr,
	    "\t-h        display help (this output)\n");
	  (void) fprintf(stderr,
	    "\t-m m      m is the macro file name\n");
	  (void) fprintf(stderr,
	    "\t-n n      n is the starting page number\n");
	  (void) fprintf(stderr,
	    "\t-o o      o specifies output pages and page ranges\n");
	  (void) fprintf(stderr,
	    "\tfile ...  source file names\n");
	  exit(Err);
	}
	if (mac == 0) {

	    /*
	     * No macros - enable Bold, Italic and Roman fonts.
	     */
		for (i = 0; Fcode[i].nm; i++) {
			switch (Fcode[i].nm) {
			case 'B':
			case 'I':
			case 'R':
				Fcode[i].status = '1';
			}
		}
	}
/*
 * Add user-supplied file names.
 */
	pc = nf;
	if (optind >= argc) {
		files[nf++] = NULL;       /* STDIN */
	} else {
		while (optind < argc) {
			if (nf >= MAXFILES)
				Error(WARN, NOLINE, " too many files at ",
					argv[optind]);
			files[nf++] = argv[optind++];
		}
	}
/*
 * Make sure all input files are accessible.
 */
	for (i = 0; i < nf; i++) {
		if (files[i] != NULL) {
			if (stat(files[i], &sbuf) != 0)
				Error(WARN, NOLINE, " can't find ", files[i]);
		}
	}
	if (Err)
		exit(1);
/*
 * Miscellaneous initialization.
 */

	for (i = 0; ; i++) {
		if (Pat[i].re == NULL)
			break;
		if ((Pat[i].pat = regcomp(Pat[i].re)) == NULL)
			Error(WARN, NOLINE, Pat[i].re, " regcomp failure");
	}
	if ((i = Findscale((int)'n', 0.0, 0)) < 0)
		Error(WARN, NOLINE, " can't find Scale['n']", NULL);
	Scalen = Scale[i].val;
	if ((i = Findscale((int)'u', 0.0, 0)) < 0)
		Error(WARN, NOLINE, " can't find Scale['u']", NULL);
	Scaleu = Scale[i].val;
	if ((i = Findscale((int)'v', 0.0, 0)) < 0)
		Error(WARN, NOLINE, " can't find Scale['v']", NULL);
	Scalev = Scale[i].val;
	(void) Findstr((unsigned char *)"CH", (unsigned char *)"= % -", 1);
	Cont = Newstr((unsigned char *)" ");
	Contlen = 1;
	if ((Trtbl = (unsigned char *)malloc(256)) == NULL)
		Error(WARN, NOLINE, " can't allocate translate table space",
			NULL);
	else {
		*Trtbl = ' ';
		for (i = 1; i < 256; i++)
			Trtbl[i] = (unsigned char) i;
	}
	if (Err)
		exit(1);
/*
 * Here begins pass1 of awf - reading input lines and expanding macros.
 */

/*
 * Output prolog.
 */
	if (Fstr.i)
		Stringput(Fstr.i, Fstr.il);
	Macro((unsigned char *)".^x");
	Macro((unsigned char *)".^b");
	Macro((unsigned char *)".^# 1 <prolog>");
/*
 * Put today's date in the -me td string or the -ms DY string.
 * Put the month name in the -ms MO string.
 */
	if (Marg == MEMACROS || Marg == MSMACROS) {
		tm = time(NULL);
		lt = localtime(&tm);
		(void) sprintf(dt, "%s %d, %d",
			Month[lt->tm_mon], lt->tm_mday, lt->tm_year + 1900);
		Findstr(
			(Marg == MEMACROS) ? (unsigned char *)"td"
					   : (unsigned char *)"DY",
			(unsigned char *)dt, 1);
		if (Marg == MSMACROS)
			Findstr((unsigned char *)"MO",
				(unsigned char *)Month[lt->tm_mon], 1);
	}
/*
 * Read input files.
 */
	for (i = 0; i < nf; i++) {
		if (files[i] == NULL) {
			np = "STDIN";
			Ifs = stdin;
		} else {

#if	defined(UNIX)
			if ((Ifs = fopen(files[i], "r")) == NULL)
#else	/* !defined(UNIX) */
			if ((Ifs = fopen(files[i], "rt")) == NULL)
#endif	/* defined(UNIX) */

				Error(FATAL, NOLINE, " can't open ", files[i]);
			np = files[i];
		}
		Findstr((unsigned char *)"IF", (unsigned char *)np, 1);
		if (i >= pc) {
			(void) sprintf((char *)Line, ".^# 1 %s", np);
			Macro(Line);
			NR = 0;
		}
		Fsp = 0;
		do {
			while (fgets((char *)Line, MAXLINE, Ifs) != NULL) {
				NR++;
				if ((np = strrchr((char *)Line, '\n')) != NULL)
					*np = '\0';
				else
					Line[MAXLINE-1] = '\0';
				Macro(Line);
			}
			if (i >= pc)
				Macro((unsigned char *)".^e");
			if (Ifs != stdin)
				(void) fclose(Ifs);
			if (Fsp > 0) {
				Free(&Inname);
				Inname = Inn_stk[Fsp-1];
				NR = NR_stk[Fsp-1];
				Ifs = Ifs_stk[Fsp-1];
			}
		} while (Fsp-- > 0);
	}
	Macro(NULL);
	if (fff)
		Charput("\f");
	return(Err);
}


/*
 * AsmPg(pa, pb) - assemble an ASCII page number
 */

char *
AsmPg(pa, pb)
	char *pa;			/* page number in ASCII */
	int *pb;			/* page number in binary */
{
	for (*pb = 0; *pa; pa++) {
		if (*pa < '0' || *pa > '9')
			return(pa);
		*pb = (*pb * 10) + (int)(*pa - '0');
	}
	return(pa);
}


/*
 * AsmPgRange(s) - assemble a page range
 */

int
AsmPgRange(s)
	char *s;			/* page range string */
{
	char *cp, *tbuf;
	int pg, pl, pu;
	struct pgrange *pr;

	if (s == NULL || (pl = strlen(s)) == 0)
		return(1);
/*
 * Remove spaces from the range string.
 */
	if ((tbuf = (char *)malloc(pl + 1)) == NULL)
		Error(FATAL, NOLINE, " no string space for page range", NULL);
	for (cp = tbuf; *s; *s++) {
		if (*s != ' ')
			*cp++ = *s;
	}
	*cp = '\0';
/*
 * Parse the range string.
 */
	for (cp = tbuf; *cp;) {

	/*
	 * Get the starting value.
	 * A missing initial range value defaults to one.
	 */
		if (*cp == '-')
			pg = 1;
		else
			cp = AsmPg(cp, &pg);
	/*
	 * Base further processing on the range terminator character.
	 */
		switch (*cp) {
		case ',':		/* single page number */
		case '\0':
			pl = pu = pg;
			break;
		case '-':		/* page number range */
			cp++;
			pl = pg;
		/*
		 * Assemble the terminating page number.  A missing
		 * number defaults to page 1,000,000 -- the end of
		 * the document.
		 */
			if (!*cp) {
				pu = 10000;
				break;
			}
			cp = AsmPg(cp, &pg);
			pu = pg;
			break;
		default:
			return(0);
		}
	/*
	 * Check the page numbers.
	 */
		if (pl < 0 || pu < 0 || pl > pu)
			return(0);
		if (*cp) {
			if (*cp != ',')
				return(0);
			cp++;
		}
	/*
	 * Allocate, fill in, and connect a page range structure.
	 */
		if ((pr = (struct pgrange *)malloc(sizeof(struct pgrange)))
		== NULL) {
			Error(FATAL, NOLINE, " no space for page range entry",
				NULL);
		}
		pr->l = pl;
		pr->u = pu;
		pr->next = PgRange;
		PgRange = pr;
	}
	return(1);
}


/*
 * Macro(inp) - process a possible macro statement
 *		pass non-macros and macros alike to pass 2
 */

void
Macro(inp)
	unsigned char *inp;		/* possible macro statement pointer */
{
	unsigned char c[2];		/* characters */
	int endm;			/* end of macro status */
	FILE *fs;			/* temporary file stream */
	int req;			/* request character status */
	unsigned char *s1, *s2;		/* temporary string pointers */

	if (inp == NULL) {
		Pass2(NULL);
		return;
	}
	req = (*inp == '.' || *inp == '\'') ? 1 : 0;
/*
 * Check for file name designator.
 */
	if (req && inp[1] == '^' && inp[2] == '#') {
		Free(&Inname);
		Inname = Field(3, inp, 1);
		F = NULL;
		Pass2(inp);
		return;
	}
/*
 * Check for source command - "^[.']so".
 */
	if (req && inp[1] == 's' && inp[2] == 'o') {
		if ((s1 = Field(2, inp, 1)) == NULL) {
			Error(WARN, LINE, " no file specified", NULL);
			return;
		}
		if ((fs = fopen((char *)s1, "r")) == NULL) {
			Error(WARN, LINE, " can't open", NULL);
			return;
		}
		if (Fsp >= MAXFSTK) {
			(void) fclose(fs);
			Error(WARN, LINE, " nesting too deep", NULL);
			return;
		}
		Ifs_stk[Fsp] = Ifs;
		Ifs = fs;
		Inn_stk[Fsp] = Inname;
		Inname = F;
		F = NULL;
		NR_stk[Fsp++] = NR;
		NR = 0;
		return;
	}
 /*
  * Check for start of macro definition.
  */
	if (req && inp[1] == 'd' && inp[2] == 'e') {
		if (inp[3] != ' ' || inp[4] == '\0') {
			Error(WARN, LINE, " illegal macro definition", NULL);
			return;
		}
		c[0] = inp[4];
		c[1] = inp[5];
		Curmx = Findmacro(c, 1);
		return;
	}
/*
 * Check for macro text.  Remove double backslashes.
 */
	if (req && (inp[1] == '\0' || (inp[2] == '\0' && inp[0] == inp[1])))
		endm = 1;
	else
		endm = 0;
	if (Curmx >= 0 && !endm) {
		if (Mtx >= MAXMTXT)
			Error(FATAL, LINE, " out of macro text space", NULL);
		if ((s1 = (unsigned char *)strchr((char *)inp, '\\')) == NULL)
			Macrotxt[Mtx] = Newstr(inp);
		else {
			for (s1 = Pass1ln, s2 = inp;; s1++) {
				if ((*s1 = *s2++) == '\0')
					break;
				if (*s1 == '\\' && *s2 == '\\')
					s2++;
			}
			Macrotxt[Mtx] = Newstr(Pass1ln);
		}
		if (Macrotab[Curmx].bx == -1)
			Macrotab[Curmx].bx = Mtx;
		Mtx++;
		Macrotab[Curmx].ct++;
		return;
	}
/*
 * Check for end of macro.
 */
	if (Curmx >= 0 && endm) {
		Curmx = -1;
		(void) sprintf((char *)Pass1ln, ".^# %d %s", NR, Inname);
		Pass2(Pass1ln);
		return;
	}
 /*
  * Check for conditionals and macro expansions.
  */
	if (req
	&&  (Findmacro(inp+1, 0) != -1 || regexec(Pat[0].pat, inp))) {
		Expand(inp);
		return;
	}
/*
 * None of the above: forward the line.
 */
	Pass2(inp);
}
